File: test/src/test/java/cz/vhromada/common/test/stub/AbstractMovableChildFacadeStub.java
package cz.vhromada.common.test.stub;

import java.util.List;

import cz.vhromada.common.Movable;
import cz.vhromada.common.converter.MovableConverter;
import cz.vhromada.common.facade.AbstractMovableChildFacade;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.validator.MovableValidator;

/**
 * A class represents stub for {@link AbstractMovableChildFacade}.
 *
 * @author Vladimir Hromada
 */
public class AbstractMovableChildFacadeStub extends AbstractMovableChildFacade<Movable, Movable, Movable, Movable> {

    /**
     * Creates a new instance of AbstractMovableChildFacadeStub.
     *
     * @param service         service for movable data
     * @param converter       converter for movable data
     * @param parentValidator validator for movable data for parent data
     * @param childValidator  validator for movable data for child data
     * @throws IllegalArgumentException if service for movable data is null
     *                                  or converter for movable data is null
     *                                  or validator for movable data for parent data is null
     *                                  or validator for movable data for child data is null
     */
    public AbstractMovableChildFacadeStub(final MovableService<Movable> service, final MovableConverter<Movable, Movable> converter,
        final MovableValidator<Movable> parentValidator, final MovableValidator<Movable> childValidator) {
        super(service, converter, parentValidator, childValidator);
    }

    @Override
    protected Movable getDomainData(final Integer id) {
        for (final Movable movable : getService().getAll()) {
            if (id.equals(movable.getId())) {
                return movable;
            }
        }

        return null;
    }

    @Override
    protected List<Movable> getDomainList(final Movable parent) {
        return List.of(getService().get(parent.getId()));
    }

    @Override
    protected Movable getForAdd(final Movable parent, final Movable data) {
        return getService().get(parent.getId());
    }

    @Override
    protected Movable getForUpdate(final Movable data) {
        return getDomainData(getDataForUpdate(data).getId());
    }

    @Override
    protected Movable getForRemove(final Movable data) {
        return getDomainData(data.getId());
    }

    @Override
    protected Movable getForDuplicate(final Movable data) {
        return getDomainData(data.getId());
    }

    @Override
    protected Movable getForMove(final Movable data, final boolean up) {
        return getDomainData(data.getId());
    }

}

File: common-test/src/test/kotlin/cz/vhromada/common/test/stub/AbstractMovableChildFacadeStub.kt
package cz.vhromada.common.test.stub

import cz.vhromada.common.domain.AuditEntity
import cz.vhromada.common.entity.Movable
import cz.vhromada.common.facade.AbstractMovableChildFacade
import cz.vhromada.common.mapper.Mapper
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.validator.MovableValidator

/**
 * A class represents stub for [AbstractMovableChildFacade].
 *
 * @author Vladimir Hromada
 */
class AbstractMovableChildFacadeStub(
        service: MovableService<AuditEntity>,
        accountProvider: AccountProvider,
        timeProvider: TimeProvider,
        mapper: Mapper<Movable, AuditEntity>,
        parentValidator: MovableValidator<Movable>,
        childValidator: MovableValidator<Movable>) : AbstractMovableChildFacade<Movable, AuditEntity, Movable, AuditEntity>(service, accountProvider, timeProvider, mapper, parentValidator, childValidator) {

    override fun getDomainData(id: Int): AuditEntity? {
        for (movable in service.getAll()) {
            if (id == movable.id) {
                return movable
            }
        }
        return null
    }

    override fun getDomainList(parent: Movable): List<AuditEntity> {
        return listOf(service.get(parent.id!!)!!)
    }

    override fun getForAdd(parent: Movable, data: AuditEntity): AuditEntity {
        return service.get(parent.id!!)!!
    }

    override fun getForUpdate(data: Movable): AuditEntity {
        val updateData = getDomainData(getDataForUpdate(data).id!!)!!
        updateData.modify(getAudit())
        return updateData
    }

    override fun getForRemove(data: Movable): AuditEntity {
        return getDomainData(data.id!!)!!
    }

    override fun getForDuplicate(data: Movable): AuditEntity {
        return getDomainData(data.id!!)!!
    }

    override fun getForMove(data: Movable, up: Boolean): AuditEntity {
        val moveData = getDomainData(data.id!!)!!
        moveData.modify(getAudit())
        return moveData
    }

}
